Modules/_io/textio.c

```diff
@@ -1498,15 +1498,26 @@ textiowrapper_read_chunk(textio *self, Py_ssize_t size_hint)
         /* Given this, we know there was a valid snapshot point
          * len(dec_buffer) bytes ago with decoder state (b'', dec_flags).
          */
-        if (PyArg_ParseTuple(state, "OO", &dec_buffer, &dec_flags) < 0) {
+        if (!PyTuple_Check(state)) {
+            PyErr_Format(PyExc_TypeError,
+                         "decoder getstate() should have returned a tuple, "
+                         "not '%.200s'",
+                         Py_TYPE(state)->tp_name);
+            Py_DECREF(state);
+            return -1;
+        }
+        if (!PyArg_ParseTuple(state,
+                              "OO;decoder getstate() should have returned a "
+                              "2-tuple", &dec_buffer, &dec_flags))
+        {
             Py_DECREF(state);
             return -1;
         }
 
         if (!PyBytes_Check(dec_buffer)) {
             PyErr_Format(PyExc_TypeError,
-                         "decoder getstate() should have returned a bytes "
-                         "object, not '%.200s'",
+                         "decoder getstate() should have returned a tuple "
```

Comment:
Good point. Similar changes should be made in `DECODER_GETSTATE()`.

I think it is worth to keep the wrong type name in this error, since returning a bytearray or a memoryview is more common error. But it would be nice to keep it short and make all three error messages similar.

For example, just "illegal decoder state" for first two cases and "illegal decoder state: the first item should be a bytes object, not '%.200s'" in the third case.